piker/brokers/kucoin.py

```diff
@@ -0,0 +1,670 @@
+    client = Client()
+    await client.cache_pairs()
+
+    yield client
+
+
+@tractor.context
+async def open_symbol_search(
+    ctx: tractor.Context,
+) -> None:
+    async with open_cached_client('kucoin') as client:
+        # load all symbols locally for fast search
+        await client.cache_pairs()
+        await ctx.started()
+
+        async with ctx.open_stream() as stream:
+            async for pattern in stream:
+                await stream.send(await client.search_symbols(pattern))
+                log.info('Kucoin symbol search opened')
+
+
+async def stream_quotes(
+    send_chan: trio.abc.SendChannel,
+    symbols: list[str],
+    feed_is_live: trio.Event,
+    loglevel: str = '',
+    # startup sync
+    task_status: TaskStatus[tuple[dict, dict]] = trio.TASK_STATUS_IGNORED,
+) -> None:
+    '''
+    Required piker api to stream real-time data.
+    Where the rubber hits the road baby
+
+    '''
+    connect_id = str(uuid4())
+
+    async with open_cached_client('kucoin') as client:
+        log.info('Starting up quote stream')
+        # loop through symbols and sub to feedz
+        for sym in symbols:
+            token, ping_interval = await client._get_ws_token()
+            pairs = await client.cache_pairs()
+            kucoin_sym = pairs[sym].symbol
+
+            init_msgs = {
+                # pass back token, and bool, signalling if we're the writer
+                # and that history has been written
+                sym: {
+                    'symbol_info': {
+                        'asset_type': 'crypto',
+                        'price_tick_size': 0.0005,
+                        'lot_tick_size': 0.1,
+                    },
+                    'shm_write_opts': {'sum_tick_vml': False},
+                    'fqsn': sym,
+                },
+            }
+            @acm
+            async def subscribe(ws: wsproto.WSConnection):
+                @acm
+                async def open_ping_task(ws: wsproto.WSConnection):
+                    '''
+                    Spawn a non-blocking task that pings the ws
+                    server every ping_interval so Kucoin doesn't drop
+                    our connection
+
+                    '''
+                    async with trio.open_nursery() as n:
+                        # TODO: cache this task so it's only called once
+                        async def ping_server():
+                            while True:
+                                await trio.sleep((ping_interval - 1000) / 1000)
+                                await ws.send_msg({'id': connect_id, 'type': 'ping'})
+
+                        log.info(f'Starting ping task for {sym}')
+                        n.start_soon(ping_server)
+
+                        yield ws
+
+                        n.cancel_scope.cancel()
+
+                # Spawn the ping task here
+                async with open_ping_task(ws) as ws:
+                    tasks = []
```

Comment:
this seems generally weird to me, the whole `tasks` bit.

i'd say just either add the subs inline without all the `.append()`ing or just generally drop these 2 kinda weird loops.

i got a feeling you can do the subs requests in maybe one send as well?

Reply:
Couldn't figure out the multiple requests in one send but I've factored out this abstraction.